### PR TITLE
8 shipment pdf client

### LINKED ISSUE #8: shpiment PDF function should be...
a collective of all packing slip tables that are included in the shipment.
Each should be its own document with header that shows the packing slip id

### PR DESCRIPTION
**Pull request by mitchell-mccracken**
Closes #8 

Solution: 
* Created `getAsPdf`, `GetOrderFulfillmentInfo`, `_pdf_makeManifestBlock`, `_pdf_makePackingBlock`, `_pdf_GetLogoURI`, `_pdf_makeBannerBlock`, and `_pdf_makeSignaturesBlock` functions to generate data structure for making shipment pdf.
* Helper functions were based off of the download helper functions used to create packing slips.

Improvements: 
* Gathering data from multiple db collections could be created with a pipeline.

### CODE REVIEW

File: src/shipment/controller.js [hunk 1]
```diff
@@ -505,3 +510,323 @@ async function getPopulatedShipmentData(shipmentId = undefined) {
...
+            }
+          }
+        }
+
+        const items = [];
+        for ( const [ ,item] of Object.entries(_report) ) {
+          items.push(item);
+        }
+
+        const pageBreakAfter =  ( orderNumbers.length === 1 || idx === (orderNumbers.length - 1) ) ?
```
Comment (jarrilla) [inline src/shipment/controller.js:573]: You don't need a ternary operator here.
if you define a variable `const x = A || B` it's going to be assigned a boolean value.

If you negate your logic, you'll accomplish the same thing you're writing in an easier to ready manner

File: src/shipment/controller.js [hunk 2]
```diff
@@ -505,3 +510,323 @@ async function getPopulatedShipmentData(shipmentId = undefined) {
...
+
+async function getAsPdf(req, res) {
+  ExpressHandler(
+    async () => {
+      const { manifest, customer, dateCreated, shipmentId, deliveryMethod } = req.body;
+      const { title } = customer;
+
+      const orderNumbers = [];
+      for ( m of manifest ) {
+        if ( !orderNumbers.includes(m.orderNumber) ) orderNumbers.push(m.orderNumber);
```
Comment (jarrilla) [inline src/shipment/controller.js:522]: This is a tad hard to read.. the same thing might be accomplished with a `Set`
for example `const orderNumbers = new Set( manifest.map(x => x.orderNumber) )`  and you can still iterate on orderNumbers as usual

File: src/shipment/controller.js [hunk 3]
```diff
@@ -505,3 +510,323 @@ async function getPopulatedShipmentData(shipmentId = undefined) {
...
+              _report[_id] = item;
+              _report[_id].qtyShipped = qty;
+            }
+            else {
+              _report[_id].qtyShipped += qty;
+            }
+          }
+        }
+
+        const items = [];
```
Comment (jarrilla) [inline src/shipment/controller.js:568]: `const items = Object.values(_report)`

File: src/shipment/controller.js [hunk 4]
```diff
@@ -505,3 +510,323 @@ async function getPopulatedShipmentData(shipmentId = undefined) {
...
+  ];
+
+  body[1].forEach((x) => {
+    x.fillColor = "#cccccc";
+    x.bold = true;
+    x.alignment = "center";
+  });
+
+  let lineNumber = 0;
+  for (let i of items) {
```
Comment (jarrilla) [inline src/shipment/controller.js:789]: Don't use `let` in a `for of` loop. It implies `i` will be manipulated, which is strongly discouraged.

File: src/shipment/controller.js (hunk 1 above)
Comment (mitchell-mccracken) [reply src/shipment/controller.js]: This makes sense. Updated

File: src/shipment/controller.js (hunk 2 above)
Comment (mitchell-mccracken) [reply src/shipment/controller.js]: Updated `orderNumbers` to be a Set. Had to add a counter ( `idx` ) to determine if `pageBreakAfter` is set to true or false.

File: src/shipment/controller.js (hunk 3 above)
Comment (mitchell-mccracken) [reply src/shipment/controller.js]: Updated

File: src/shipment/controller.js (hunk 4 above)
Comment (mitchell-mccracken) [reply src/shipment/controller.js]: Understood

**Review by jarrilla — APPROVED** [review]

**Pull request merged**